handlers/whatsapp/whatsapp.go

```diff
@@ -523,38 +533,119 @@ func (h *handler) SendMsg(ctx context.Context, msg courier.Msg) (courier.MsgStat
 	return status, nil
 }
 
-func sendWhatsAppMsg(msg courier.Msg, url string, token string, payload interface{}) (string, *courier.ChannelLog, error) {
+func sendWhatsAppMsg(msg courier.Msg, sendPath *url.URL, token string, payload interface{}) (string, []*courier.ChannelLog, error) {
+	start := time.Now()
 	jsonBody, err := json.Marshal(payload)
+
 	if err != nil {
-		log := courier.NewChannelLog("unable to build JSON body", msg.Channel(), msg.ID(), "", "", courier.NilStatusCode, "", "", time.Duration(0), err)
-		return "", log, err
+		elapsed := time.Now().Sub(start)
+		log := courier.NewChannelLogFromError("unable to build JSON body", msg.Channel(), msg.ID(), elapsed, err)
+		return "", []*courier.ChannelLog{log}, err
 	}
-
-	req, _ := http.NewRequest(http.MethodPost, url, bytes.NewReader(jsonBody))
-	req.Header.Set("Content-Type", "application/json")
-	req.Header.Set("Accept", "application/json")
-	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", token))
-	req.Header.Set("User-Agent", utils.HTTPUserAgent)
+	req, _ := http.NewRequest(http.MethodPost, sendPath.String(), bytes.NewReader(jsonBody))
+	req.Header = buildWhatsAppRequestHeader(token)
 	rr, err := utils.MakeHTTPRequest(req)
 	log := courier.NewChannelLogFromRR("Message Sent", msg.Channel(), msg.ID(), rr).WithError("Message Send Error", err)
-	if err != nil {
-		return "", log, err
+	errPayload := &mtErrorPayload{}
+
+	// handle send msg errors
+	if err = json.Unmarshal(rr.Body, errPayload); err == nil && len(errPayload.Errors) > 0 {
```

Comment:
Please unpack this. For the super simple cases this construct is ok but for anything more than the simple != `nil` case it should be exploded.

Reply:
Hmm... doesn't the `len(errPayload.Errors) > 0` statement ensure that the errors will be handled safely?

Reply:
All I'm saying here is unpack it.. ie:

```golang
err = json.Unmarshal(rr.Body, errPayload);
if err == nil && len(errPayload.Errors) > 0 {
   ...
```